Issue/5

HwYun opened the pull request.

HwYun committed:
로그인 AOP 기능 구현

HwYun committed:
UserController login 메소드 반환값 유저 정보를 포함하게 변경
및 피드백 적용

HwYun committed:
Resolve Conflict

ccommit-dev committed:
Merge pull request #15 from j-lab-edu/issue/4

Issue/4

HwYun committed:
암호화 라이브러리 Spring Security에서 JBCrypt로변경

HwYun committed:
Product CRUD 개발

HwYun committed:
User CRUD 리팩토링 및 오류 수정

HwYun committed:
UserLevel에 따른 권한 범위 조정

j-lab-dev commented on shop-server/build.gradle:
```diff
@@ -25,11 +25,15 @@ dependencies {
 
 	// MyBatis
-	implementation 'org.mybatis.spring.boot:mybatis-spring-boot-starter:2.2.0'
+	implementation 'org.mybatis.spring.boot:mybatis-spring-boot-starter:2.2.2'
```
minor 버전이 바뀌긴 했지만 변경사항 또는 변경한 이유를 간략하게 주석으로 추가해주시는게 좋을것같습니다.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/aop/LoginCheck.java:
```diff
@@ -0,0 +1,16 @@
+@Target(ElementType.METHOD)
+public @interface LoginCheck {
+    enum UserLevel {
+        USER, SELLER, ADMIN
```
USER 하면 판매도 가능한것 처럼 보여서
GENRALUSER, BUIER 같이 명시적으로 바꾸는게 좋을것 같습니다.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/aop/LoginCheckAspect.java:
```diff
@@ -0,0 +1,78 @@
+                parameters[i] = userId;
+            }
+        }
+        return parameters;
```
파라미터값이 비어있을때  예외상황을 추가해주시면 좋을것 같습니다.
1) AOP(assertJ문법)으로 예외상황을 필터한다.
2) Controller 예외상황에 대한 분기문을 추가한다.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/util/BcryptEncoder.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class BcryptEncoder {
```
BcryptEncoder 암호화 알고리즘 내부 동작 원리를 주석으로 위에 추가해주시면 좋을것 같습니다.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/service/impl/ProductServiceImpl.java:
```diff
@@ -1,25 +1,54 @@
-    public void updateProducts(ProductDTO productDTO){
+    @Override
+    public ProductDTO register(String userId, ProductDTO productDTO) {
+        int accountId = userMapper.getAccountId(userId);
```
uuid 의 정의, 용도, 필요성을 찾아서 댓글에 남겨주세요.

j-lab-dev replied:
AutoIncrement의 필요성도 같이 찾아주세요.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/service/impl/ProductServiceImpl.java:
```diff
@@ -1,25 +1,54 @@
+        productDTO.setUpdateDate(new Date());
+        productDTO.setReviewCount(0);
+        productDTO.setTotalStarRating(0);
+        productDTO.setPurchaseCount(0);
```
db 필드 추가시 default 명시로 설정해두는 방법도 있습니다.

j-lab-dev commented on shop-server/src/main/resources/mapper/productMapper.xml:
```diff
@@ -0,0 +1,40 @@
+    </insert>
+
+    <select id="getMyProducts" resultType="com.dev.shopserver.dto.ProductDTO">
+        SELECT *
```
모든 필드를 명시적으로 작성해주시면 운영시에 가독성이 올라가 유지보수하기 좋은 코드가 됩니다.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/dto/UserDTO.java:
```diff
@@ -2,14 +2,18 @@
 
 import lombok.Getter;
 import lombok.Setter;
+import org.apache.catalina.User;
```
불필요한 패키지는 삭제해주세요.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/dto/UserDTO.java:
```diff
@@ -2,14 +2,18 @@
 public class UserDTO {
     public enum Status {
-        DEFAULT, ADMIN, SELLER, DELETED
+        DEFAULT,
```
GENERAL로 명칭을 바꿔주시면 좋을것 같습니다.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/dto/UserDTO.java:
```diff
@@ -2,14 +2,18 @@
+        DEFAULT,
+        ADMIN,
+        SELLER,
+        DELETED;
```
유저 로그 Table (히스토리성) 추가 후에 
삭제 api 호출마다 로그 table 에 날짜 유저 정보를 기록하고 삭제하시면 될것 같습니다.

j-lab-dev commented on shop-server/src/main/resources/mapper/userMapper.xml:
```diff
@@ -25,13 +19,14 @@
-               phone,
-               address,
-               status
+        SELECT *
```
필드 명시 부탁드립니다.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/aop/LoginCheckAspect.java:
```diff
@@ -37,19 +39,27 @@ public Object checkLogin(ProceedingJoinPoint proceedingJoinPoint, LoginCheck log
         switch (userLevel){
             case "USER": {
-                userId = (String)session.getAttribute("LOGIN_USER_ID");
+                userId = (String)session.getAttribute(key);
```
user의 레벨 변경시에 세션 clear 가 필요할것 같습니다.

j-lab-dev commented on shop-server/src/main/java/com/dev/shopserver/aop/LoginCheckAspect.java:
```diff
@@ -37,19 +39,27 @@ public Object checkLogin(ProceedingJoinPoint proceedingJoinPoint, LoginCheck log
             }
             case "SELLER": {
-                userId = (String)session.getAttribute("LOGIN_SELLER_ID");
+                if(!key.equals("LOGIN_USER_ID")) userId = (String) session.getAttribute(key);
```
지금은 유저 레벨 상태가 3가지라
status 값이 단일이어도 위같은 분기로 처리 가능하겠지만

상태가 계속 추가된다면 확장성있는 코드는 아닌거같습니다.
상태를 list로 갖고있어서 list안에 index접근 후 해당 권한이 있는지 비교하는 코드로 변경해주세요.

HwYun commented on shop-server/src/main/resources/mapper/productMapper.xml:
```diff
@@ -0,0 +1,40 @@
+
+    <update id="updateProduct" parameterType="com.dev.shopserver.dto.ProductDTO">
+        UPDATE product
+        SET price = #{price},
```
필드의 값이 유효한(0, null, "") 이아닌 값만 왔을때 set이 되게 mybatis 문법 중 <switch> 문을 추가해주세요

HwYun replied:
<choose>

HwYun replied:
"<chose>"

HwYun replied:
https://mybatis.org/mybatis-3/ko/dynamic-sql.html

HwYun committed:
Fix: 유저 권한체크 및 로그인 방식 변경

유저 레벨별로 세션에 다르게 저장하지 않고 같은 방식으로 저장
현재 로그인한 유저의 ID와 Status를 세션에 저장하도록 변경
jackson 및 Mybatis 설정의 비효율성으로 인해 UserDTO의 Status를 Enum에서 단순한 String으로 변경

ccommit-dev reviewed (commented):
머지 완료하시면 될것같습니다.

HwYun committed:
Refactor: mapper SQL 쿼리 수정

userMapper.xml 은 가독성 좋게 수정
productMapper.xml 은 update 쿼리에 null 체크 추가
null 체크를 위해서 productDTO 필드를 Wrapper 클래스로 변경
